src/Host/Broker/Impl/Pipes/MessagePipe.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.Extensions.Logging;
+using Microsoft.R.Host.Client;
+
+namespace Microsoft.R.Host.Broker.Pipes {
+    public class MessagePipe {
+        private readonly ILogger _logger;
+        private readonly BufferBlock<byte[]> _hostMessages = new BufferBlock<byte[]>();
+        private readonly BufferBlock<byte[]> _clientMessages = new BufferBlock<byte[]>();
+
+        private ConcurrentDictionary<ulong, byte[]> _sentPendingRequests = new ConcurrentDictionary<ulong, byte[]>();
+        private Queue<byte[]> _unsentPendingRequests = new Queue<byte[]>();
+
+        private byte[] _handshake;
+        private IMessagePipeEnd _hostEnd, _clientEnd;
+
+        private abstract class PipeEnd : IMessagePipeEnd {
```

Comment:
In this case it looks like `HostEnd` and `MessagePipe` should be combined into one - `HostEnd` is just a wrapper around `MessagePipe` methods. Also, I'd create `IMessagePipeClientEnd : IMessagePipeEnd, IDisposable` to remove 
```{csharp}
            public override void Dispose() {
                throw new InvalidOperationException("Host end of the pipe should not be disposed.");
            }
```

---
In reply to: [74838114](https://github.com/Microsoft/RTVS/pull/2173#discussion_r74838114) [](ancestors = 74838114,74837833,74833662,74833429)